File: c++/file_structure/sort/insertion_sort.cc
#include "sort.h"
using namespace std;

std::vector<int> sort::insertion_sort(std::vector<int> &input) {
    std::vector<int> ret =
        std::vector<int>(input.begin(), input.end()); // copy input
    for (int i = 2; (long unsigned int)i <= ret.size(); ++i) {
        int target = ret[i - 1];
        for (int j = i - 1; j > 0; --j) {
            if (ret[j - 1] > target) {
                ret[j] = ret[j - 1];
                ret[j - 1] = target;
            } else {
                break; // j is not incremented when breaking
            }
        }
    }
    return ret;
}

File: rust/file_structure/src/sort/insertion_sort.rs
pub fn insertion_sort(input: &[i64]) -> Vec<i64> {
    let mut ret = input.to_vec();
    for i in 2..=ret.len() {
        let key = ret[i - 1];
        let mut j = i - 1;
        while j > 0 && ret[j - 1] > key {
            ret[j] = ret[j - 1];
            j -= 1;
        }
        ret[j] = key;
    }
    ret
}
